tracer/src/Datadog.Trace/Configuration/ConfigurationSources/Telemetry/ConfigurationBuilder.cs

```diff
@@ -0,0 +1,300 @@
+
+            public int? Get(Func<int, bool> validator) => Get(null, validator);
+
+            [return: NotNullIfNotNull(nameof(defaultValue))] // This doesn't work with nullables, but it still expresses intent
+            public int? Get(int? defaultValue, Func<int, bool>? validator)
+            {
+                var result = _keys.Source.GetInt32(_keys.Key, _keys.Telemetry, validator)
+                          ?? (_keys.FallbackKey1 is null ? null : _keys.Source.GetInt32(_keys.FallbackKey1, _keys.Telemetry, validator))
+                          ?? (_keys.FallbackKey2 is null ? null : _keys.Source.GetInt32(_keys.FallbackKey2, _keys.Telemetry, validator))
+                          ?? (_keys.FallbackKey3 is null ? null : _keys.Source.GetInt32(_keys.FallbackKey3, _keys.Telemetry, validator));
+
+                // We have a valid value
+                if (result is { Result: { } value, IsValid: true })
+                {
+                    return value;
+                }
+
+                // don't have a default value
+                if (defaultValue is null)
+                {
+                    return null;
+                }
+
+                _keys.Telemetry.Record(_keys.Key, defaultValue.Value, ConfigurationOrigins.Default);
+                return defaultValue.Value;
+            }
+        }
+
+        internal readonly struct DictionaryAccessor
+        {
+            private readonly HasKeys _keys;
+
+            public DictionaryAccessor(HasKeys keys)
+            {
+                _keys = keys;
+            }
+
+            public IDictionary<string, string>? Get() => Get(allowOptionalMappings: false);
+
+            public IDictionary<string, string>? Get(bool allowOptionalMappings)
+            {
+                // TODO: Handle/allow default values + validation?
+                var result = _keys.Source.GetDictionary(_keys.Key, _keys.Telemetry, validator: null, allowOptionalMappings)
+                          ?? (_keys.FallbackKey1 is null ? null : _keys.Source.GetDictionary(_keys.FallbackKey1, _keys.Telemetry, validator: null, allowOptionalMappings))
+                          ?? (_keys.FallbackKey2 is null ? null : _keys.Source.GetDictionary(_keys.FallbackKey2, _keys.Telemetry, validator: null, allowOptionalMappings))
+                          ?? (_keys.FallbackKey3 is null ? null : _keys.Source.GetDictionary(_keys.FallbackKey3, _keys.Telemetry, validator: null, allowOptionalMappings));
+
+                // We have a valid value
+                if (result is { Result: { } value, IsValid: true })
+                {
+                    return value;
+                }
+
+                // Horrible that we have to stringify the dictionary, but that's all that's available in the telemetry api
+                // _keys.Telemetry.Record(_keys.Key, string.Join(", ", defaultValue.Select(kvp => $"{kvp.Key}:{kvp.Value}")), ConfigurationOrigins.Default);
+                // return defaultValue;
```

Comment:
> Is the comment on purpose, if yes why

I haven't added all the potential overloads to the `DictionaryAccessor` type. In particular, there's currently no way of providing a "default" value to use. That's because:

- We currently only ever use an empty dictionary as a "default" value
- It didn't seem worth recording an empty dictionary in config
- It's a PITA to serialize the dictionary (as described in the comment)
- It makes the `AsDictionary()` overloads more complicated

I left the comment there as a hint for implementers _in case_ we do need to add a non-empty dictionary default value. We can remove it.